File: go/operators.go
package amino

import "sort"

// OperatorFunc implements an operator on the in-process target. Prefix
// operators receive one argument, infix two.
type OperatorFunc func(args ...any) (any, error)

// OperatorDef registers an operator with the parser. Exactly one of Symbol or
// Keyword is set. InputTypes and ReturnType drive type checking; a "*" input
// type accepts anything. Fn may be nil for and/or/not, which the evaluator
// implements itself.
type OperatorDef struct {
	Symbol        string
	Keyword       string
	Kind          string // "infix" or "prefix"
	Fn            OperatorFunc
	BindingPower  int
	Associativity string // "left" or "right"
	InputTypes    []string
	ReturnType    string
}

// Token returns the text the operator is written as.
func (d *OperatorDef) Token() string {
	if d.Symbol != "" {
		return d.Symbol
	}
	return d.Keyword
}

type operatorRegistry struct {
	byToken  map[string][]*OperatorDef
	symbols  map[string]bool
	keywords map[string]bool
}

func newOperatorRegistry() *operatorRegistry {
	return &operatorRegistry{byToken: map[string][]*OperatorDef{}, symbols: map[string]bool{}, keywords: map[string]bool{}}
}

func (r *operatorRegistry) register(d OperatorDef) error {
	if (d.Symbol == "") == (d.Keyword == "") {
		return newError(CodeConfig, "operator needs exactly one of Symbol or Keyword")
	}
	if d.Kind == "" {
		d.Kind = "infix"
	}
	if d.Kind != "infix" && d.Kind != "prefix" {
		return newError(CodeConfig, "operator %q: kind must be infix or prefix, got %q", d.Token(), d.Kind)
	}
	if d.Associativity == "" {
		d.Associativity = "left"
	}
	if d.InputTypes == nil {
		if d.Kind == "prefix" {
			d.InputTypes = []string{"*"}
		} else {
			d.InputTypes = []string{"*", "*"}
		}
	}
	if d.ReturnType == "" {
		d.ReturnType = "Bool"
	}
	tok := d.Token()
	for _, existing := range r.byToken[tok] {
		if sameTypes(existing.InputTypes, d.InputTypes) {
			return newError(CodeOperatorConflict, "operator %q with input types %v already registered", tok, d.InputTypes)
		}
	}
	def := d
	r.byToken[tok] = append(r.byToken[tok], &def)
	if d.Symbol != "" {
		r.symbols[d.Symbol] = true
	} else {
		r.keywords[d.Keyword] = true
	}
	return nil
}

func sameTypes(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func (r *operatorRegistry) candidates(tok string) []*OperatorDef { return r.byToken[tok] }

func (r *operatorRegistry) bindingPower(tok string) (int, bool) {
	c := r.byToken[tok]
	if len(c) == 0 {
		return 0, false
	}
	return c[0].BindingPower, true
}

func (r *operatorRegistry) prefix(tok string) *OperatorDef {
	for _, d := range r.byToken[tok] {
		if d.Kind == "prefix" {
			return d
		}
	}
	return nil
}

func (r *operatorRegistry) symbolList() []string {
	out := make([]string, 0, len(r.symbols))
	for s := range r.symbols {
		out = append(out, s)
	}
	sort.Slice(out, func(i, j int) bool { return len(out[i]) > len(out[j]) })
	return out
}

// Built-in operators. and/or/not are the irreducible minimum and are always
// present. Their InputTypes are what the type checker enforces.
var alwaysOperators = map[string]bool{"or": true, "and": true, "not": true}

func standardOperators() []OperatorDef {
	cmp := func(f func(int) bool) OperatorFunc {
		return func(args ...any) (any, error) {
			c, err := compareValues(args[0], args[1])
			if err != nil {
				return nil, err
			}
			return f(c), nil
		}
	}
	return []OperatorDef{
		{Keyword: "or", BindingPower: 10, InputTypes: []string{"Bool", "Bool"}},
		{Keyword: "and", BindingPower: 20, InputTypes: []string{"Bool", "Bool"}},
		{Keyword: "not", Kind: "prefix", BindingPower: 30, InputTypes: []string{"Bool"}},
		{Keyword: "in", BindingPower: 40, InputTypes: []string{"*", "List"}, Fn: func(args ...any) (any, error) {
			return inList(args[0], args[1])
		}},
		{Keyword: "not in", BindingPower: 40, InputTypes: []string{"*", "List"}, Fn: func(args ...any) (any, error) {
			in, err := inList(args[0], args[1])
			if err != nil {
				return nil, err
			}
			return !in, nil
		}},
		{Symbol: "=", BindingPower: 40, Fn: func(args ...any) (any, error) { return equalValues(args[0], args[1]) }},
		{Symbol: "!=", BindingPower: 40, Fn: func(args ...any) (any, error) {
			eq, err := equalValues(args[0], args[1])
			if err != nil {
				return nil, err
			}
			return !eq, nil
		}},
		{Symbol: ">", BindingPower: 40, Fn: cmp(func(c int) bool { return c > 0 })},
		{Symbol: "<", BindingPower: 40, Fn: cmp(func(c int) bool { return c < 0 })},
		{Symbol: ">=", BindingPower: 40, Fn: cmp(func(c int) bool { return c >= 0 })},
		{Symbol: "<=", BindingPower: 40, Fn: cmp(func(c int) bool { return c <= 0 })},
		{Keyword: "contains", BindingPower: 40, InputTypes: []string{"Str", "Str"}, Fn: func(args ...any) (any, error) {
			hay, ok1 := args[0].(string)
			needle, ok2 := args[1].(string)
			if !ok1 || !ok2 {
				return nil, newError(CodeEvaluation, "contains needs two strings")
			}
			return containsString(hay, needle), nil
		}},
	}
}

func buildOperatorRegistry(preset string, names []string) (*operatorRegistry, error) {
	all := standardOperators()
	var chosen []OperatorDef
	switch {
	case names != nil:
		enabled := map[string]bool{}
		for _, n := range names {
			enabled[n] = true
		}
		for _, d := range all {
			if enabled[d.Token()] || alwaysOperators[d.Token()] {
				chosen = append(chosen, d)
			}
		}
	case preset == "" || preset == "standard":
		chosen = all
	case preset == "minimal":
		for _, d := range all {
			if alwaysOperators[d.Token()] {
				chosen = append(chosen, d)
			}
		}
	default:
		return nil, newError(CodeConfig, "unknown operator preset %q", preset)
	}
	r := newOperatorRegistry()
	for _, d := range chosen {
		if err := r.register(d); err != nil {
			return nil, err
		}
	}
	return r, nil
}

File: typescript/src/operators.ts
import { AminoError } from "./errors.js";
import { compareValues, equalValues, inList } from "./values.js";

/** Implements an operator on the in-process target. Prefix operators get one argument, infix two. */
export type OperatorFn = (...args: unknown[]) => unknown;

/**
 * Registers an operator with the parser. Exactly one of symbol or keyword is
 * set. inputTypes and returnType drive type checking; "*" accepts anything.
 * fn may be omitted for and/or/not, which the evaluator implements itself.
 */
export interface OperatorDef {
  symbol?: string;
  keyword?: string;
  kind?: "infix" | "prefix";
  fn?: OperatorFn;
  bindingPower: number;
  associativity?: "left" | "right";
  inputTypes?: string[];
  returnType?: string;
}

/** A fully defaulted operator as stored in the registry. */
export interface ResolvedOperator {
  token: string;
  kind: "infix" | "prefix";
  fn: OperatorFn | undefined;
  bindingPower: number;
  associativity: "left" | "right";
  inputTypes: string[];
  returnType: string;
  isSymbol: boolean;
}

export class OperatorRegistry {
  private readonly byToken = new Map<string, ResolvedOperator[]>();

  register(def: OperatorDef): void {
    const hasSymbol = def.symbol !== undefined && def.symbol !== "";
    const hasKeyword = def.keyword !== undefined && def.keyword !== "";
    if (hasSymbol === hasKeyword) throw new AminoError("config", "operator needs exactly one of symbol or keyword");
    const kind = def.kind ?? "infix";
    const op: ResolvedOperator = {
      token: (hasSymbol ? def.symbol : def.keyword) as string,
      kind,
      fn: def.fn,
      bindingPower: def.bindingPower,
      associativity: def.associativity ?? "left",
      inputTypes: def.inputTypes ?? (kind === "prefix" ? ["*"] : ["*", "*"]),
      returnType: def.returnType ?? "Bool",
      isSymbol: hasSymbol,
    };
    const existing = this.byToken.get(op.token) ?? [];
    for (const e of existing) {
      if (e.inputTypes.length === op.inputTypes.length && e.inputTypes.every((t, i) => t === op.inputTypes[i])) {
        throw new AminoError("operator_conflict", `operator "${op.token}" with input types [${op.inputTypes.join(", ")}] already registered`);
      }
    }
    existing.push(op);
    this.byToken.set(op.token, existing);
  }

  candidates(token: string): ResolvedOperator[] {
    return this.byToken.get(token) ?? [];
  }

  bindingPower(token: string): number | undefined {
    return this.byToken.get(token)?.[0]?.bindingPower;
  }

  prefix(token: string): ResolvedOperator | undefined {
    return this.byToken.get(token)?.find((d) => d.kind === "prefix");
  }

  /** Symbol tokens, longest first, for the tokenizer. */
  symbols(): string[] {
    const out: string[] = [];
    for (const [tok, defs] of this.byToken) if (defs[0]?.isSymbol) out.push(tok);
    return out.sort((a, b) => b.length - a.length);
  }
}

export const ALWAYS = new Set(["or", "and", "not"]);

export function standardOperators(): OperatorDef[] {
  const cmp =
    (f: (c: number) => boolean): OperatorFn =>
    (a, b) =>
      f(compareValues(a, b));
  return [
    { keyword: "or", bindingPower: 10, inputTypes: ["Bool", "Bool"] },
    { keyword: "and", bindingPower: 20, inputTypes: ["Bool", "Bool"] },
    { keyword: "not", kind: "prefix", bindingPower: 30, inputTypes: ["Bool"] },
    { keyword: "in", bindingPower: 40, inputTypes: ["*", "List"], fn: (a, b) => inList(a, b) },
    { keyword: "not in", bindingPower: 40, inputTypes: ["*", "List"], fn: (a, b) => !inList(a, b) },
    { symbol: "=", bindingPower: 40, fn: (a, b) => equalValues(a, b) },
    { symbol: "!=", bindingPower: 40, fn: (a, b) => !equalValues(a, b) },
    { symbol: ">", bindingPower: 40, fn: cmp((c) => c > 0) },
    { symbol: "<", bindingPower: 40, fn: cmp((c) => c < 0) },
    { symbol: ">=", bindingPower: 40, fn: cmp((c) => c >= 0) },
    { symbol: "<=", bindingPower: 40, fn: cmp((c) => c <= 0) },
    {
      keyword: "contains",
      bindingPower: 40,
      inputTypes: ["Str", "Str"],
      fn: (a, b) => {
        if (typeof a !== "string" || typeof b !== "string") throw new AminoError("evaluation", "contains needs two strings");
        return a.includes(b);
      },
    },
  ];
}

export function builtinTokens(): Set<string> {
  return new Set(standardOperators().map((d) => (d.symbol ?? d.keyword) as string));
}

export function buildOperatorRegistry(preset: "standard" | "minimal" | string[] | undefined): OperatorRegistry {
  const all = standardOperators();
  let chosen: OperatorDef[];
  if (Array.isArray(preset)) {
    const enabled = new Set(preset);
    chosen = all.filter((d) => enabled.has((d.symbol ?? d.keyword) as string) || ALWAYS.has(d.keyword ?? ""));
  } else if (preset === undefined || preset === "standard") {
    chosen = all;
  } else if (preset === "minimal") {
    chosen = all.filter((d) => ALWAYS.has(d.keyword ?? ""));
  } else {
    throw new AminoError("config", `unknown operator preset "${String(preset)}"`);
  }
  const r = new OperatorRegistry();
  for (const d of chosen) r.register(d);
  return r;
}
